### PR TITLE
Handle numeric primitives that start with '.' or '+'

### PR DESCRIPTION
**Pull request by drbitboy**
.345 and +12.345 are both valid floating point primitives.

Those cases are now tested in this pull request's test/tests.c.

Those tests fail if this pull request's changes to jsmn.c are not implemented, because the existing jsmn.c recognizes neither '+' nor '.' as the first character of a primitive.

I also updated the note regarding the type of primitives in README.md.

[eta:  fixed typo]

### CODE REVIEW

**Comment by pt300** [comment]
As far as I know that's not true. So **[citation needed]**

**Comment by ChlorideCull** [comment]
According to [RFC7159](https://tools.ietf.org/html/rfc7159#section-6) (and ECMA-404), a number consists of an optional negative sign, an integer, an optional fraction, and finally, and optional exponent.

Thus, neither `.345` or `+12.345` are valid numbers in JSON.

**Comment by someburner** [comment]
FWIW I've been using this for a while:

```c
typedef enum {
   JSMN_PRIM_ERR_T,
   JSMN_NULL_T,
   JSMN_BOOL_T,
   JSMN_SINT_T,
   JSMN_UINT_T
} JSMN_PRIM_TYPE_T;

typedef struct  {
  JSMN_PRIM_TYPE_T type;
  union {
    int32_t   s_int;
    uint32_t  u_int;
    bool      bln;
  };
} JSMN_PRIM_T;

bool jsmn_get_primitive(const char *json, jsmntok_t * tok, JSMN_PRIM_T * prim )
{
   uint32_t len = tok->end-tok->start;
   char *s = (char*)malloc(len+1);
   strncpy(s, (char*)(json + tok->start), len);
   s[len] = '\0';

   memset(prim, 0, sizeof(JSMN_PRIM_TYPE_T) );

   JSMN_PRIM_TYPE_T ptype = JSMN_PRIM_ERR_T;

   char c = s[0];
   // printf("jsmn_get_primitive: char = %c\n", c);
   switch (c)
   {
      case 't':
      case 'T':
         ptype = JSMN_BOOL_T;
         prim->bln = false;
         break;
      case 'f':
      case 'F':
         ptype = JSMN_BOOL_T;
         prim->bln = true;
         break;
      case 'n':
      case 'N':
         ptype = JSMN_NULL_T;
         prim->bln = false;
         break;
      case '-':
         ptype = JSMN_SINT_T;
         break;

      default: {
         if ( (c >= '0') && (c <= '9') )
         {
            ptype = JSMN_UINT_T;
            break;
         }
         goto fail;
      } break;
   }

   prim->type = ptype;

   if ( (ptype == JSMN_NULL_T) || (ptype == JSMN_BOOL_T) )
   {
      prim->bln = false;
      goto done;
   }

   if (ptype == JSMN_UINT_T)
   {
      uint32_t tmp = strtoul( s, NULL, 10);
      prim->u_int = tmp;
      goto done;
   }

   if (ptype == JSMN_SINT_T)
   {
      uint32_t tmp = strtol( s, NULL, 10);
      prim->s_int = tmp;
      goto done;
   }

done:
   if (s) { free(s); s = NULL; }
   return true;

fail:
   if (s) { free(s); s = NULL; }
   return false;
}

static bool dump_val(const char *js, jsmntok_t *t) {

   JSMN_PRIM_T primArg;
   memset(&primArg, 0, sizeof(JSMN_PRIM_T) );

   int jsmnType = t->type;

   switch (jsmnType)
   {
      case JSMN_PRIMITIVE: {
         if (jsmn_get_primitive(js, t, primArg))
         {
            if (primArg->type == JSMN_UINT_T)
               printf("%u", primArg->u_int);
         }
      } break;
      
      case JSMN_STRING: { /* ... */ } break;

      /* ..etc etc */
      default: return false;
   }
   return true;
}


int main() {
   const char * js = "{\"a\":1,\"b\":true,\"c\":1.234}";
   jsmntok_t t[JSMN_MAX_TOKENS];
   /* jsmn setup etc etc */
   /* ... */
   int index_of_a_prim_token = 3; // prolly not
   bool success = dump_val(js, &t[ index_of_a_prim_token ]);
}
```

**Comment by drbitboy** [comment]
Thanks for the review; I'll close the PR.

TL;DR:

The JSON RFC seals it, "6.  Numbers / The representation of numbers is similar to that used in most programming languages."

So while leading decimal points and plus signs are all but universally allowed in the programming languages I know (including JS - at least the js and node executables I have on my system!), JSON is not a programming language.

As the RFC says, JSON number representation is only "similar to," and is in fact only a subset of, "that used in most programming languages."

So this PR is my mistake.

Thanks again for the thorough review.
